### PR TITLE
Update README, process, and proposal template for Swift 4 stage 1.

### PR DESCRIPTION
**Pull request by DougGregor**

### CODE REVIEW

File: 0000-template.md [hunk 1]
```diff
@@ -44,13 +44,55 @@ comments detailing what it does. The detail in this section should be
...
-possible to migrate existing Swift code to use a new feature or API
-automatically?
+## Source compatibility
+
+Relative to the Swift 3 evolution process, the source compatibility
+requirements for Swift 4 are *much* more stringent: we should only
+break source compatibility if the Swift 3 constructs were actively
+harmful in some way, the volume of affected Swift 3 code is relatively
+small, and we can provide source compatibility (in Swift 3
+compatiblity mode) and migration.
```
Comment (lattner) [inline 0000-template.md]: typo compatiblity

File: README.md [hunk 2]
```diff
@@ -20,20 +20,78 @@ historical purposes, but are not necessarily indicative of the
...
+  fundamental types in the language. Swift 4 seeks to make strings more
+  powerful and easier-to-use, while retaining Unicode correctness by
+  default.
+
+* Memory ownership model: an (opt-in) Cyclone/Rust-inspired memory
+  ownership model is highly desired by systems programmers and for
+  other high-performance applications that want predictible and
+  deterministic performance. This feature will fundamentally shape the
+  ABI, from low-level language concerns such as "inout" and low-level
+  "addressors" to its impact on the standard library.
```
Comment (lattner) [inline README.md]: I'd suggest mentioning that we need to get a design of this, which will allow us to scope the small part that effects the ABI and make sure that happens in Stage 1.

Comment (lattner) [inline README.md]: IOW, ownership as a whole is not a stage 1 feature.

File: README.md [hunk 3]
```diff
@@ -20,20 +20,78 @@ historical purposes, but are not necessarily indicative of the
...
+
+## Swift 3.0 - Released on September 13, 2016
+
```
Comment (lattner) [inline README.md]: Would it make sense to break descriptions of historical releases out to a separate page?  We could have a "Swift 3" page, which lists its raison d'etre, along with the SE's that went into it.

**Comment by lattner** [comment]
This looks really great Doug, I added a few comments to the patch.

File: 0000-template.md [hunk 4]
```diff
@@ -44,13 +44,55 @@ comments detailing what it does. The detail in this section should be
...
+aid in migration?
+
+## Effect on ABI stability
+
+Does the proposal change the ABI of existing language features? The
+ABI comprises all aspects of the code generation model and interaction
+with the Swift runtime, including such things as calling conventions,
+the layout of data types, and the behavior of dynamic features in the
+language (reflection, dynamic dispatch, dynamic casting via `as?`,
+etc.). Purely syntactic changes rarely affect existing ABI. Additive
```
Comment (MrBendel) [inline 0000-template.md]: Most of the time, you'll want affect as a verb meaning to influence something and effect for the something that was influenced. The difference between affect and effect is so slippery that people have started using "impact" as a verb instead. Don't be one of them!

File: README.md [hunk 5]
```diff
@@ -20,20 +20,78 @@ historical purposes, but are not necessarily indicative of the
...
+Expected release date: Late 2017
+
+The Swift 4 release is designed around two primary goals: to provide
+source stability for Swift 3 code and to provide ABI stability for the
+Swift standard library. To that end, the Swift 4 release will be
+divided into two stages.
+
+Stage 1 focuses on the essentials required for source and ABI
+stability. Features that don't fundamentally change the ABI of
+existing language features or imply and ABI-breaking change to the
```
Comment (milseman) [inline README.md]: ... or imply *an* ABI-breaking ...

File: 0000-template.md [hunk 6]
```diff
@@ -44,13 +44,55 @@ comments detailing what it does. The detail in this section should be
...
+requirements for Swift 4 are *much* more stringent: we should only
+break source compatibility if the Swift 3 constructs were actively
+harmful in some way, the volume of affected Swift 3 code is relatively
+small, and we can provide source compatibility (in Swift 3
+compatiblity mode) and migration.
+
+Will existing correct Swift 3 or Swift 4 applications stop compiling
+due to this change? Will applications still compile but produce
+different behavior than they used to? If "yes" to either of these, is
+it possible for the Swift 4 compiler to accept the old syntax in it's
```
Comment (jpsim) [inline 0000-template.md]: "in it's" -> "in its"

File: 0000-template.md (hunk 4 above)
Comment (jtbandes) [inline 0000-template.md]: "effect *on* ABI" ⇒ "affect [the] ABI". 😄

Comment (DougGregor) [inline 0000-template.md]: I'll just strike the words "affect" and "effect" from the English language and save us all some time.

File: README.md (hunk 2 above)
Comment (DougGregor) [inline README.md]: Sure.

File: README.md (hunk 5 above)
Comment (DougGregor) [inline README.md]: thx

File: README.md (hunk 3 above)
Comment (DougGregor) [inline README.md]: Yeah, I like that better. It keeps the README shorter and more forward-looking.

File: 0000-template.md (hunk 4 above)
Comment (jtbandes) [inline 0000-template.md]: https://xkcd.com/326/

**Comment by lattner** [comment]
Nice, LGTM!  Thank you for working on this.

File: README.md [hunk 7]
```diff
@@ -20,94 +20,80 @@ historical purposes, but are not necessarily indicative of the
...
+  library.
+
+* String re-evaluation: String is one of the most important
+  fundamental types in the language. Swift 4 seeks to make strings more
+  powerful and easier-to-use, while retaining Unicode correctness by
+  default.
+
+* Memory ownership model: an (opt-in) Cyclone/Rust-inspired memory
+  ownership model is highly desired by systems programmers and for
+  other high-performance applications that want predictible and
```
Comment (schneiderandre) [inline README.md]: "predictible" -> "predictable"

**Pull request merged**